lightning/src/util/ser.rs

```diff
@@ -510,11 +511,72 @@ impl_array!(3); // for rgb
 impl_array!(4); // for IPv4
 impl_array!(12); // for OnionV2
 impl_array!(16); // for IPv6
-impl_array!(32); // for channel id & hmac
+impl_array!(32); // for channel id, hmac, and XOnlyPublicKey
 impl_array!(PUBLIC_KEY_SIZE); // for PublicKey
-impl_array!(COMPACT_SIGNATURE_SIZE); // for Signature
+impl_array!(64); // for ecdsa::Signature and schnorr::Signature
 impl_array!(1300); // for OnionPacket.hop_data
 
+/// For variable-length values within TLV subtypes where the length cannot be inferred from the
+/// enclosing TLV record.
+///
+/// The length is encoded as an unsigned `U` type.
+#[derive(Clone)]
+pub(crate) struct WithLength<T, U>(pub T, pub core::marker::PhantomData<U>);
+
+impl<T: Writeable> Writeable for WithLength<Vec<T>, u8> {
+	#[inline]
+	fn write<W: Writer>(&self, w: &mut W) -> Result<(), io::Error> {
+		(self.0.len() as u8).write(w)?;
+		for element in self.0.iter() {
+			element.write(w)?;
+		}
+		Ok(())
+	}
+}
+impl<T: Readable> Readable for WithLength<Vec<T>, u8> {
+	#[inline]
+	fn read<R: Read>(r: &mut R) -> Result<Self, DecodeError> {
+		let len = <u8 as Readable>::read(r)? as usize;
+		let mut result = Vec::with_capacity(len);
+		for _ in 0..len {
+			result.push(<T as Readable>::read(r)?);
+		}
+		Ok(Self(result, core::marker::PhantomData))
+	}
+}
+
+/// For variable-length values within TLV record where the length is encoded as part of the record.
+/// Used to prevent encoding the length twice.
+#[derive(Clone)]
+pub(crate) struct WithoutLength<T>(pub T);
```

Comment:
Ok, so I whipped together a macro that is compliant with the spec. However, if a `vec_type` field is missing when parsing, it will be later serialized as present but with no values (i.e., a parsing-serialization roundtrip results in a different serialization).

Reply:
> Then we shouldn't write something like that in an `optional_vec` and reject it on the read end?

I guess I'm just not sure if it is technically invalid.